_posts/2019-12-24-release-r118/index.md

```diff
@@ -0,0 +1,207 @@
+
+This Snowplow beta release includes the long-awaited new bad row format. We have updated the microservices that make up the Snowplow pipeline to emit much more highly structured bad rows, that should be significantly easier to query and debug: making it simpler for companies running Snowplow to realibly monitor for data quality issues, spot when they arise, and then quickly diagnose the source of type of problem tha that has emerged, the event or entity that it impacts, the size of the problem and the app it has occurred on. For more information on the thinking behind this new functionality, see the [original RFC][RFC].
+
+In addition, this release also includes important refactoring in the libraries that we use (e.g. `cats` instead of `scalaz` and `circe` instead of`json4s`),
+a bump of [Beam](https://beam.apache.org/) version to 2.11.0,
+as well as an improvement to the referer parser enrichment.
+
+[Scala Stream Collector](https://github.com/snowplow/snowplow/tree/master/2-collectors/scala-stream-collector)
+and the enrich jobs versions ([Stream Enrich](https://github.com/snowplow/snowplow/tree/master/3-enrich/stream-enrich),
+[Beam Enrich](https://github.com/snowplow/snowplow/tree/master/3-enrich/beam-enrich),
+and [Spark Enrich](https://github.com/snowplow/snowplow/tree/master/3-enrich/spark-enrich)) are bumped to `1.0.0`.
+
+1. [Beta release](#betarelease)
+2. [New bad row format](#badrows)
+3. [Referer parser enrichment improvement](#refererparser)
+4. [Upgrading](#upgrading)
+5. [Roadmap](#roadmap)
+6. [Getting help](#help)
+
+<h1 id="betarelease">1. Beta release</h1>
+
+This release represents a major upgrade of the core Snowplow technology. We are currently testing this tech internally - no small feat because of the size and scope of the release. Whilst we are running those tests, we have decided to make the code available to open source users to test themselves: we would welcome any feedback on the new functionality, which we hope to release in production in [R119](#roadmap).
+
+If you are an open source users who would like to test this new functionality, We would be happy to provide a high-priority OSS support during this testing phase (covering setup, debug, recovery or other aspects). We are committed to iterating our release candidates for R119 rapidly.
+
+<h1 id="badrows">2. New bad row format</h1>
+
+<h2>2.1. Why update the bad row format?</h2>
+
+The Snowplow pipeline has always been non-lossy: when there is an issue processing an event, rather than silently discard the event, it is saved as a "bad row" - opening up the possibility to monitor the number of bad rows, inspect them, fix them and reprocess them.
+
+The previous format for these bad rows (generic JSON) was not straightforward to use. Specifically:
+
+1. It was hard to distinguish "bad rows" that mattered i.e. that were real data that was not successfully processed, with rubbish data e.g. generated by bots crawling the collector endpoint
```

Comment:
Not sure if "POST bombs" problem is the part of this issue or separate one, but here's one event in a POST payload contains an error - old bad row would give the whole payload, making it hard to find problematic rows and opening a DDoS attack vector.

Reply:
I mentioned it in the part about enrichment failures:
```
For those familiar with the previous bad rows emitted in case of enrichment failure, please note that with this new version, when an error occurs while enriching an event (that can come from a collector payload with several events), the bad row will contain only this event. The other events of the collector payload can be successfully enriched. This is the same for schema violations.
```